x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/ProfilingDataStreamManager.java

```diff
@@ -0,0 +1,238 @@
+        );
+        dataStreams.add(ProfilingDataStream.of("profiling-metrics", 1));
+        dataStreams.add(ProfilingDataStream.of("profiling-hosts", 1));
+        PROFILING_DATASTREAMS = Collections.unmodifiableList(dataStreams);
+    }
+
+    private final ThreadPool threadPool;
+    private final Client client;
+
+    public ProfilingDataStreamManager(ThreadPool threadPool, Client client, ClusterService clusterService) {
+        super(clusterService);
+        this.threadPool = threadPool;
+        this.client = client;
+    }
+
+    @Override
+    protected void onStatus(
+        ClusterState clusterState,
+        Status status,
+        ProfilingDataStream index,
+        ActionListener<? super ActionResponse> listener
+    ) {
+        switch (status) {
+            case NEEDS_CREATION -> createDataStream(index, listener);
+            case NEEDS_VERSION_BUMP -> rolloverDataStream(index, listener);
+            default -> {
+                logger.debug("Skipping status change [{}] for data stream [{}].", status, index);
+                // ensure that listener is notified we're done
+                listener.onResponse(null);
+            }
+        }
+    }
+
+    protected IndexMetadata indexMetadata(ClusterState state, ProfilingDataStream dataStream) {
+        Map<String, DataStream> dataStreams = state.metadata().dataStreams();
+        if (dataStreams == null) {
+            return null;
+        }
+        DataStream ds = dataStreams.get(dataStream.getName());
+        if (ds == null) {
+            return null;
+        }
+        Index writeIndex = ds.getWriteIndex();
+        if (writeIndex == null) {
+            return null;
+        }
+        return state.metadata().index(writeIndex);
+    }
+
+    @Override
+    protected Iterable<ProfilingDataStream> getManagedIndices() {
+        return PROFILING_DATASTREAMS;
+    }
+
+    private void onDataStreamFailure(ProfilingDataStream dataStream, Exception ex) {
+        logger.error(() -> format("error for data stream [%s] for [%s]", dataStream, ClientHelper.PROFILING_ORIGIN), ex);
+    }
+
+    private void rolloverDataStream(final ProfilingDataStream dataStream, ActionListener<? super ActionResponse> listener) {
+        logger.debug("rolling over data stream [{}].", dataStream);
+        final Executor executor = threadPool.generic();
+        executor.execute(() -> {
+            RolloverRequest request = new RolloverRequest(dataStream.getName(), null);
+            request.masterNodeTimeout(TimeValue.timeValueMinutes(1));
+            executeAsyncWithOrigin(
+                client.threadPool().getThreadContext(),
+                ClientHelper.PROFILING_ORIGIN,
+                request,
+                new ActionListener<RolloverResponse>() {
+                    @Override
+                    public void onResponse(RolloverResponse response) {
+                        if (response.isAcknowledged() == false) {
```

Comment:
just for readability: is it possible to move the logging code into a separate function, e.g. logResponseStatus(...) ? Same for `rolloverIndex()` in ProfilingIndexManager.